source/common/http/conn_pool_base.cc

```diff
@@ -1,16 +1,51 @@
 #include "common/http/conn_pool_base.h"
 
+#include "common/network/transport_socket_options_impl.h"
+#include "common/runtime/runtime_features.h"
 #include "common/stats/timespan_impl.h"
 #include "common/upstream/upstream_impl.h"
 
 namespace Envoy {
 namespace Http {
+Network::TransportSocketOptionsSharedPtr
+wrapTransportSocketOptions(Network::TransportSocketOptionsSharedPtr transport_socket_options,
+                           Protocol protocol) {
+  if (!Runtime::runtimeFeatureEnabled("envoy.reloadable_features.http_default_alpn")) {
+    return transport_socket_options;
+  }
+
+  // If configured to do so, we override the ALPN to use for the upstream connection to match the
+  // selected protocol.
+  std::vector<std::string> alpn;
+  switch (protocol) {
+  case Http::Protocol::Http10:
+    NOT_REACHED_GCOVR_EXCL_LINE;
+  case Http::Protocol::Http11:
+    alpn.push_back("http/1.1");
+    break;
+  case Http::Protocol::Http2:
+    alpn.push_back("h2");
+    break;
+  case Http::Protocol::Http3:
+    NOT_REACHED_GCOVR_EXCL_LINE;
```

Comment:
nit: not implemented as I think this will eventually be needed? Also add a TODO?